config/import-control-test.xml

```diff
@@ -19,18 +19,42 @@
 
   <!-- Conflicts with normal tests and pitest.
        See examples in https://github.com/checkstyle/checkstyle/issues/6439 -->
-  <allow class="org.powermock.reflect.Whitebox" />
   <allow class="org.mockito.internal.util.Checks" />
   <disallow pkg="org\.powermock.*" regex="true" />
   <disallow pkg="org\.mockito.*" regex="true" />
+  <!-- Reflection shouldn't be used in tests. -->
+  <disallow pkg="java\.lang\.reflect\.*" regex="true" />
```

Comment:
Why is reflection not allowed part of the JDK 17 compatibility but then we allow it in other places in the import control? Are those tests that use reflect not JDK 17 compliant? Is there more work to make checkstyle JDK 17 compliant?

This disallow is caught up in #9142 and because of the HUGE suppression list at the time there remains more reflections areas with possibly no plan to correct them since they are not connected to #9142.

Example:
https://github.com/checkstyle/checkstyle/blob/272dbe394e1d581aec6743a7c7eba0aab0d61f9b/src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java#L42-L43
`CheckerTest` is now completely Truth compliant in my future PRs, but I get violations when removing the suppression because this class uses reflection.

Reply:
The full list is:
````
[ERROR] \api\JavadocTokenTypesTest.java:25:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \CheckerTest.java:38:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \CheckerTest.java:39:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\coding\RequireThisCheckTest.java:28:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \checks\header\HeaderCheckTest.java:32:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \checks\imports\CustomImportOrderCheckTest.java:31:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\imports\ImportControlLoaderTest.java:32:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \checks\imports\ImportControlLoaderTest.java:33:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\javadoc\JavadocMethodCheckTest.java:30:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \checks\javadoc\JavadocMethodCheckTest.java:31:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\javadoc\JavadocTagInfoTest.java:24:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\SuppressWarningsHolderTest.java:25:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \checks\SuppressWarningsHolderTest.java:26:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \checks\SuppressWarningsHolderTest.java:27:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \checks\TranslationCheckTest.java:33:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \checks\UniquePropertiesCheckTest.java:29:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \checks\UniquePropertiesCheckTest.java:30:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \ConfigurationLoaderTest.java:29:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \ConfigurationLoaderTest.java:30:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \ConfigurationLoaderTest.java:31:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \DefaultLoggerTest.java:28:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \DefaultLoggerTest.java:29:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \DetailAstImplTest.java:28:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \grammar\GeneratedJavaTokenTypesTest.java:24:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \grammar\GeneratedJavaTokenTypesTest.java:25:1: Disallowed import - java.lang.reflect.Modifier. [ImportControlTest]
[ERROR] \grammar\javadoc\GeneratedJavadocTokenTypesTest.java:24:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \grammar\javadoc\GeneratedJavadocTokenTypesTest.java:25:1: Disallowed import - java.lang.reflect.Modifier. [ImportControlTest]
[ERROR] \internal\AllChecksTest.java:24:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \internal\AllChecksTest.java:25:1: Disallowed import - java.lang.reflect.Modifier. [ImportControlTest]
[ERROR] \internal\XdocsPagesTest.java:33:1: Disallowed import - java.lang.reflect.Array. [ImportControlTest]
[ERROR] \internal\XdocsPagesTest.java:34:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \internal\XdocsPagesTest.java:35:1: Disallowed import - java.lang.reflect.ParameterizedType. [ImportControlTest]
[ERROR] \MainTest.java:36:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \MainTest.java:37:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \PackageObjectFactoryTest.java:36:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
[ERROR] \PackageObjectFactoryTest.java:37:1: Disallowed import - java.lang.reflect.Method. [ImportControlTest]
[ERROR] \PropertyCacheFileTest.java:33:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \utils\AnnotationUtilTest.java:25:1: Disallowed import - java.lang.reflect.InvocationTargetException. [ImportControlTest]
[ERROR] \utils\CommonUtilTest.java:40:1: Disallowed import - java.lang.reflect.Constructor. [ImportControlTest]
[ERROR] \utils\TokenUtilTest.java:25:1: Disallowed import - java.lang.reflect.Field. [ImportControlTest]
````

Reply:
> Why is reflection not allowed part of the JDK 17 compatibility but then we allow it in other places in the import control?

This should be a dedicated issue. All reflection-based test should be reviewed and changed (or whitelisted).
For example, https://github.com/checkstyle/checkstyle/blob/272dbe394e1d581aec6743a7c7eba0aab0d61f9b/src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java#L42-L43 should be rewritten with `TestUtil#getInternalState` method.

Reply:
https://github.com/checkstyle/checkstyle/issues/11123